python/ray/experimental/sgd/tf/tf_runner.py

```diff
@@ -83,48 +146,119 @@ def setup_distributed(self, urls, world_rank, world_size):
         with self.strategy.scope():
             self.model = self.model_creator(self.config)
 
-        # For use in model.evaluate()
-        self.local_model = None
+        self._LambdaCallback = _load_proper_LambdaCallback()
 
-    def step(self):
-        """Runs a training epoch and updates the model parameters."""
-        fit_default_config = {"verbose": self.verbose}
-        fit_default_config.update(self.config.get("fit_config", {}))
+    # runs on another thread
+    def _record_progress(self, batchN, logs):
+        self._step_counter += 1
+        self._logs = logs
 
-        history = self.model.fit(self.train_dataset, **fit_default_config)
-        if history is None:
-            stats = {}
-        else:
-            stats = {"train_" + k: v[-1] for k, v in history.history.items()}
+    # runs on another thread
+    def _has_next_batch(self, batchN, logs):
+        if batchN != 0:
+            # always gate the first batch so the model doesn't immediately run
+            if self._step_counter < self._step_limit:
+                return
+
+            self._recorded_all_results.set()
 
+        # there is a choice here of whether to run
+        # the model thread while waiting for the next step call
+        # solved by moving the barrier to _record_logs to stop
+        # training after it ends rather than before it begins
+        self._ready_to_continue.wait()
+        self._step_counter = 0
+
+    def _generic_step(self, number_of_steps, config_key, action_fn, res_fn):
```

Comment:
Can you document this method? It'll be very important to document the specific implementation for future reference.